source/extensions/filters/http/on_demand/on_demand_update.cc

```diff
@@ -4,16 +4,41 @@
 #include "common/common/enum_to_int.h"
 #include "common/common/logger.h"
 #include "common/http/codes.h"
+#include "common/upstream/cds_api_impl.h"
+
+#include "extensions/filters/http/well_known_names.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace OnDemand {
 
+OnDemandFilterConfig::OnDemandFilterConfig(
+    const envoy::extensions::filters::http::on_demand::v3::OnDemand& proto_config,
+    Upstream::ClusterManager& cm, Stats::Scope& scope,
+    ProtobufMessage::ValidationVisitor& validation_visitor)
+    : cm_(cm) {
+  if (proto_config.has_odcds_config()) {
+    odcds_ =
```

Comment:
This creates an instance of CDS api/subscription in a worker thread, which is different from the rest of xDS that runs on the main thread. I think this is an issue -- there are concurrency (definitely with ADS configured) and synchronisation issues here. I probably missed it -- where is the CDS subscription started?

Reply:
This is a good question. I'm still in progress of writing the tests, so this will come out soon. But yeah, looks like creating CDS in worker thread calls into cluster manager directly, so looks like I either would need to make it thread safe, or postpone the creation of cds until I'm in main thread.

Basically I didn't want to start the subscription, because it will result in periodically requesting the management server, which is not what I wanted. I'll have a look what VHDS is doing here.

Reply:
We did a bit of tracking where this is called from and I have an impression that both `createRouteSpecificFilterConfigTyped` and `createFilterFactoryFromProtoTyped` are called from the main thread. And inside those functions we create the CDS, so it ought to be fine. Maybe the factories that those functions return are called from worker threads, aren't they?

Reply:
> Basically I didn't want to start the subscription, because it will result in periodically requesting the management server, which is not what I wanted. I'll have a look what VHDS is doing here.

A subscription is basically type-specific (i.e. CDS, RDS, VHDS, etc) grpc connection to a management server; in case of ADS all subscriptions are multiplexed over a single grpc connection. The connection needs to be initiated (subscription started) in order to be able to both send and receive xDS updates.

Note that VHDS establishes a separate subscription in addition to the base RDS one and calculates a common state (based on RDS state and VHDS updates). I encourage you to take a closer look at various xDS implementations, as well as muxes. Please also take a look at the docs I linked to earlier.